cmd/clusterctl/client/repository/repository_memory.go

```diff
@@ -0,0 +1,172 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package repository
+
+import (
+	"fmt"
+
+	"github.com/pkg/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/serializer"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1alpha4"
+	clusterctlv1 "sigs.k8s.io/cluster-api/cmd/clusterctl/api/v1alpha3"
+	"sigs.k8s.io/cluster-api/cmd/clusterctl/internal/scheme"
+)
+
+// MemoryRepository contains an instance of the repository data.
+type MemoryRepository struct {
+	defaultVersion string
+	rootPath       string
+	componentsPath string
+	versions       map[string]bool
+	files          map[string][]byte
+}
+
+var _ Repository = &MemoryRepository{}
+
+// NewMemoryRepository returns a new MemoryRepository instance.
+func NewMemoryRepository() *MemoryRepository {
+	return &MemoryRepository{
+		versions: map[string]bool{},
+		files:    map[string][]byte{},
+	}
+}
+
+// DefaultVersion returns the default version for this repository.
+// NOTE: The DefaultVersion is a required info usually derived from the repository configuration,
+// and it is used whenever the users gets files from the repository without providing a specific version.
+func (f *MemoryRepository) DefaultVersion() string {
+	if f.defaultVersion == "" {
+		return latestVersionTag
+	}
+	return f.defaultVersion
+}
+
+// RootPath returns the RootPath for this repository.
+// NOTE: The RootPath is a required info usually derived from the repository configuration,
+// and it is used to map the file path to the internal repository structure.
+func (f *MemoryRepository) RootPath() string {
+	return f.rootPath
+}
+
+// ComponentsPath returns ComponentsPath for this repository
+// NOTE: The ComponentsPath is a required info usually derived from the repository configuration,
+// and it is used to identify the components yaml for the provider.
+func (f *MemoryRepository) ComponentsPath() string {
+	return f.componentsPath
+}
+
+// GetFile returns a file for a given provider version.
+// NOTE: If the provided version is missing, a new one will be generated automatically.
```

Comment:
I'm probably missing context but isn't it rather `If the provided version is missing, the default version is used.` ?

Reply:
Or I'm just missing where we generate the version.

Reply:
you are correct, i think a wrong copy/paste